providers/Spotify/mod.ts

```diff
@@ -0,0 +1,321 @@
+		return {
+			accessToken: content?.access_token,
+			validUntilTimestamp: Date.now() + (content.expires_in * 1000),
+		};
+	}
+}
+
+export class SpotifyReleaseLookup extends ReleaseApiLookup<SpotifyProvider, Album> {
+	constructReleaseApiUrl(): URL {
+		let lookupUrl: URL;
+		const query = new URLSearchParams();
+		if (this.lookup.method === 'gtin') {
+			lookupUrl = new URL(`search`, this.provider.apiBaseUrl);
+			query.set('type', 'album');
+			query.set('q', `upc:${this.lookup.value}`);
+		} else { // if (method === 'id')
+			lookupUrl = new URL(`albums/${this.lookup.value}`, this.provider.apiBaseUrl);
+		}
+
+		lookupUrl.search = query.toString();
+		return lookupUrl;
+	}
+
+	protected async getRawRelease(): Promise<Album> {
+		if (this.lookup.method === 'gtin') {
+			// Spotify does not always find UPC barcodes but expects them prefixed with
+			// 0 to a length of 14 characters. E.g. "810121774182" gives no results,
+			// but "00810121774182" does.
+			let albumId: string | undefined;
+			while (true) {
+				const cacheEntry = await this.provider.query<SearchResult>(
+					this.constructReleaseApiUrl(),
+					this.options.snapshotMaxTimestamp,
+				);
+				if (cacheEntry.content?.albums?.items?.length) {
+					albumId = cacheEntry.content.albums.items[0].id;
+					break;
+				} else if (this.lookup.value.length < 14) {
+					// Prefix the GTIN with 0s
+					this.lookup.value = this.lookup.value.padStart(14, '0');
+				} else {
+					// No results found
+					break;
+				}
+			}
+
+			// No results found
+			if (!albumId) {
+				throw new ResponseError(this.provider.name, 'API returned no results', this.constructReleaseApiUrl());
+			}
+
+			// Result is a SimplifiedAlbum. Perform a regular ID lookup with the found release
+			// ID to retrieve complete data.
+			this.lookup.method = 'id';
+			this.lookup.value = albumId;
+		}
+
+		const cacheEntry = await this.provider.query<Album>(
+			this.constructReleaseApiUrl(),
+			this.options.snapshotMaxTimestamp,
+		);
+		const release = cacheEntry.content;
+
+		this.updateCacheTime(cacheEntry.timestamp);
+		return release;
+	}
+
+	private async getRawTracklist(rawRelease: Album): Promise<Track[]> {
+		const allTracks: SimplifiedTrack[] = [...rawRelease.tracks.items];
+
+		// The initial response contains max. 50 tracks. Fetch the remaining
+		// tracks with separate requests if needed.
+		let nextUrl = rawRelease.tracks.next;
+		while (nextUrl && allTracks.length < rawRelease.tracks.total) {
+			const cacheEntry = await this.provider.query<ResultList<SimplifiedTrack>>(
+				new URL(nextUrl),
+				this.options.snapshotMaxTimestamp,
+			);
+			this.updateCacheTime(cacheEntry.timestamp);
+			allTracks.push(...cacheEntry.content.items);
+			nextUrl = cacheEntry.content.next;
+		}
+
+		// Load full details including ISRCs
+		if (this.options.withISRC) {
+			return this.getRawTrackDetails(allTracks);
+		} else {
+			return allTracks.map((track) => track as Track);
```

Comment:
I don't think so, because then we don't have `track.external_ids?.isrc` available later. I think it is better to be more explicit with the type in the function interfaces further down.

That's why I changed it from `SimplifiedTrack` to `Track` in 7549841c9470b2dbedb1cd69d676fcb50c5887bf

Reply:
Ok, I saw your comments below. I adjusted it to do the type conversion only when accessing the ISRC

Reply:
Yeah, GitHub's way to display reviews is bit confusing. I think it displays review comments in the order they were created and not in the order they appear in the code.
It looked more meaningful when I submitted the review, but I see that this comment (which I edited a few times) is not very helpful without the other comments below appearing next to it.